src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -156,21 +158,44 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
 
         async Task WatchKubernetesResource<T>(Func<WatchEventType, T, Task> handler) where T : CustomResource
         {
-            try
+            var retryUntilCancelled = new RetryStrategyOptions()
             {
-                await foreach (var (eventType, resource) in kubernetesService.WatchAsync<T>(cancellationToken: cancellationToken))
+                ShouldHandle = new PredicateBuilder().HandleInner<EndOfStreamException>(),
+                BackoffType = DelayBackoffType.Constant,
+                MaxRetryAttempts = int.MaxValue,
+                UseJitter = true,
+                MaxDelay = TimeSpan.FromSeconds(30),
+                OnRetry = (retry) =>
                 {
-                    await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                    _logger.LogDebug(
+                        retry.Outcome.Exception,
+                        "Long poll watch operation was ended by server after {LongPollDurationInMs} milliseconds.",
+                        retry.Duration.TotalMilliseconds
+                        );
+                    return ValueTask.CompletedTask;
+                }
+            };
 
-                    try
-                    {
-                        await handler(eventType, resource).ConfigureAwait(false);
-                    }
-                    finally
+            var pipeline = new ResiliencePipelineBuilder().AddRetry(retryUntilCancelled).Build();
+
+            try
+            {
+                await pipeline.ExecuteAsync(async (pipelineCancellationToken) =>
+                {
+                    await foreach (var (eventType, resource) in kubernetesService.WatchAsync<T>(cancellationToken: pipelineCancellationToken))
                     {
-                        semaphore.Release();
+                        await semaphore.WaitAsync(pipelineCancellationToken).ConfigureAwait(false);
+
+                        try
+                        {
+                            await handler(eventType, resource).ConfigureAwait(false);
+                        }
+                        finally
+                        {
+                            semaphore.Release();
+                        }
                     }
-                }
+                }, cancellationToken).ConfigureAwait(false);
             }
             catch (Exception ex) when (ex is not OperationCanceledException)
```

Comment:
We preserve existing behavior where we swallow all exceptions and log an error except for cancellations which bubble out.